### PR TITLE
Hods 233 : Addee CtUtr ans SaUtr for Corporations and accompanying tests. 

### PR DESCRIPTION
**Pull request by parkeradam**

### CODE REVIEW

File: app/uk/gov/hmrc/testuser/services/Generator.scala [hunk 1]
```diff
@@ -65,23 +65,23 @@ class Generator @Inject()(val testUserRepository: TestUserRepository, val config
...
-      nino                <- whenF(NATIONAL_INSURANCE,MTD_INCOME_TAX)(generateNino)
-      mtdItId             <- whenF(MTD_INCOME_TAX)(generateMtdId)
-      eoriNumber          <- whenF(CUSTOMS_SERVICES, CTC, GOODS_VEHICLE_MOVEMENTS)(useProvidedOrGenerateEoriNumber(eoriNumber))
-      vrn                 <- whenF(MTD_VAT)(generateVrn)
-      vatRegistrationDate = vrn.map(_ => LocalDate.now.minusYears(Gen.chooseNum(1,20).sample.get))
-      groupIdentifier     = Some(generateGroupIdentifier)
-      individualDetails   = generateIndividualDetails
-      userFullName        = generateUserFullName(individualDetails.firstName, individualDetails.lastName)
-      emailAddress        = generateEmailAddress(individualDetails.firstName, individualDetails.lastName)
+      saUtr <- whenF(SELF_ASSESSMENT)(generateSaUtr)
```
Comment (ajchilts) [inline app/uk/gov/hmrc/testuser/services/Generator.scala:75]: Any reason we have changed the formatting here? I prefer it all lined up?

File: app/uk/gov/hmrc/testuser/services/Generator.scala [hunk 2]
```diff
@@ -101,25 +101,26 @@ class Generator @Inject()(val testUserRepository: TestUserRepository, val config
...
     def when[T](keys: ServiceKey*)(thenDo: => T): Option[T] = Generator.when(services)(keys)(thenDo)
 
     for {
-      saUtr           <- whenF(SELF_ASSESSMENT)(generateSaUtr)
-      nino            <- whenF(NATIONAL_INSURANCE, MTD_INCOME_TAX)(generateNino)
-      mtdItId         <- whenF(MTD_INCOME_TAX)(generateMtdId)
-      empRef          <- whenF(PAYE_FOR_EMPLOYERS)(generateEmpRef)
-      ctUtr           <- whenF(CORPORATION_TAX)(generateCtUtr)
-      vrn             <- whenF(SUBMIT_VAT_RETURNS,MTD_VAT)(generateVrn)
+      saUtr <- whenF(SELF_ASSESSMENT)(generateSaUtr)
```
Comment (ajchilts) [inline app/uk/gov/hmrc/testuser/services/Generator.scala:108]: As above

File: app/uk/gov/hmrc/testuser/services/Generator.scala [hunk 3]
```diff
@@ -101,25 +101,26 @@ class Generator @Inject()(val testUserRepository: TestUserRepository, val config
...
+      lisaManRefNum <- whenF(LISA)(generateLisaManRefNum)
+      setRefNum = when(SECURE_ELECTRONIC_TRANSFER)(generateSetRefNum)
+      psaId = when(RELIEF_AT_SOURCE)(generatePsaId)
+      eoriNumber <- whenF(CUSTOMS_SERVICES, CTC, SAFETY_AND_SECURITY, GOODS_VEHICLE_MOVEMENTS)(useProvidedOrGenerateEoriNumber(eoriNumber))
       groupIdentifier = Some(generateGroupIdentifier)
-      firstName       = generateFirstName
-      lastName        = generateLastName
-      userFullName    = generateUserFullName(firstName, lastName)
-      emailAddress    = generateEmailAddress(firstName, lastName)
+      firstName = generateFirstName
```
Comment (ajchilts) [inline app/uk/gov/hmrc/testuser/services/Generator.scala:120]: As above

File: app/uk/gov/hmrc/testuser/services/Generator.scala [hunk 4]
```diff
@@ -195,43 +198,64 @@ class Generator @Inject()(val testUserRepository: TestUserRepository, val config
...
-  private def generateEmpRef: Future[String] = generateUniqueIdentifier(() => { employerReferenceGenerator.sample.get.toString })
-  private def generateSaUtr: Future[String] = generateUniqueIdentifier(() => { utrGenerator.next })
-  private def generateNino: Future[String] = generateUniqueIdentifier(() => { ninoGenerator.nextNino.value })
-  private def generateCtUtr: Future[String] = generateUniqueIdentifier(() => { utrGenerator.next })
-  private def generateVrn: Future[String] = generateUniqueIdentifier(() => { Vrn(vrnGenerator.sample.get.toString).vrn })
-  private def generateLisaManRefNum: Future[String] = generateUniqueIdentifier(() => { lisaManRefNumGenerator.next.lisaManagerReferenceNumber })
-  private def generateMtdId: Future[String] = generateUniqueIdentifier(() => { mtdItIdGenerator.next.mtdItId })
-  private def generateEoriNumber: Future[String] = generateUniqueIdentifier(() => { eoriGenerator.sample.get.value })
+  private def generateEmpRef: Future[String] = generateUniqueIdentifier(() => {
+    employerReferenceGenerator.sample.get.toString
```
Comment (ajchilts) [inline app/uk/gov/hmrc/testuser/services/Generator.scala:202]: As above. Why the change to the formatting?

File: app/uk/gov/hmrc/testuser/services/Generator.scala [hunk 5]
```diff
@@ -195,43 +198,64 @@ class Generator @Inject()(val testUserRepository: TestUserRepository, val config
...
+  })
 }
 
 class UtrGenerator(random: Random = new Random) extends Modulus11Check {
-  def this(seed: Int) = this(new scala.util.Random(seed))
-
   def next: String = {
     val randomCode = f"${random.nextInt(1000000000)}%09d"
-    val checkCharacter  = calculateCheckCharacter(randomCode)
+    val checkCharacter = calculateCheckCharacter(randomCode)
```
Comment (ajchilts) [inline app/uk/gov/hmrc/testuser/services/Generator.scala:245]: As above

File: app/uk/gov/hmrc/testuser/services/Generator.scala [hunk 6]
```diff
@@ -195,43 +198,64 @@ class Generator @Inject()(val testUserRepository: TestUserRepository, val config
...
   }
 }
 
 class ArnGenerator(random: Random = new Random) extends Modulus23Check {
-  def this(seed: Int) = this(new scala.util.Random(seed))
-
   def next: String = {
     val randomCode = "ARN" + f"${random.nextInt(1000000)}%07d"
-    val checkCharacter  = calculateCheckCharacter(randomCode)
+    val checkCharacter = calculateCheckCharacter(randomCode)
```
Comment (ajchilts) [inline app/uk/gov/hmrc/testuser/services/Generator.scala:253]: As above

File: app/uk/gov/hmrc/testuser/services/Generator.scala [hunk 7]
```diff
@@ -195,43 +198,64 @@ class Generator @Inject()(val testUserRepository: TestUserRepository, val config
...
     s"$checkCharacter$randomCode"
   }
 }
 
 class MtdItIdGenerator(random: Random = new Random) extends Modulus23Check {
-  def this(seed: Int) = this(new scala.util.Random(seed))
```
Comment (ajchilts) [inline app/uk/gov/hmrc/testuser/services/Generator.scala:233]: Why have we deleted this?

File: app/uk/gov/hmrc/testuser/services/Generator.scala [hunk 8]
```diff
@@ -240,17 +264,13 @@ class MtdItIdGenerator(random: Random = new Random) extends Modulus23Check {
 }
 
 class LisaGenerator(random: Random = new Random) extends Modulus23Check {
-  def this(seed: Int) = this(new scala.util.Random(seed))
```
Comment (ajchilts) [inline app/uk/gov/hmrc/testuser/services/Generator.scala:243]: Why have we deleted this?

File: app/uk/gov/hmrc/testuser/services/Generator.scala [hunk 9]
```diff
@@ -240,17 +264,13 @@ class MtdItIdGenerator(random: Random = new Random) extends Modulus23Check {
...
-
   def next: LisaManagerReferenceNumber = {
     val randomCode = if (random.nextBoolean()) f"${random.nextInt(999999)}%06d" else f"${random.nextInt(9999)}%04d"
     LisaManagerReferenceNumber(s"Z$randomCode")
   }
 }
 
 class SecureElectronicTransferReferenceNumberGenerator(random: Random = new Random) {
-  def this(seed: Int) = this(new scala.util.Random(seed))
```
Comment (ajchilts) [inline app/uk/gov/hmrc/testuser/services/Generator.scala:252]: Why have we deleted this?

File: app/uk/gov/hmrc/testuser/services/Generator.scala [hunk 10]
```diff
@@ -267,15 +287,15 @@ class SecureElectronicTransferReferenceNumberGenerator(random: Random = new Rand
...
-    val initialDigit = random.nextInt(9) + 1    //bug random.nextInt(8) -> random.nextInt(9)
+    val initialDigit = random.nextInt(9) + 1 //bug random.nextInt(8) -> random.nextInt(9)
     val remainingDigits = f"${random.nextInt(Int.MaxValue)}%011d"
     s"$initialDigit$remainingDigits"
   }
 }
 
 
 class PensionSchemeAdministratorIdentifierGenerator(random: Random = new Random) {
-  def this(seed: Int) = this(new scala.util.Random(seed))
```
Comment (ajchilts) [inline app/uk/gov/hmrc/testuser/services/Generator.scala:278]: Why have we deleted this?

**Review by ajchilts — CHANGES_REQUESTED** [review]

**Review by Luka-Abey — APPROVED** [review]

**Review by ajchilts — APPROVED** [review]

**Comment by ajchilts** [comment]
Fantastic work @parkeradam :-)

File: conf/app.routes [hunk 11]
```diff
@@ -16,3 +16,6 @@ GET         /individuals/vrn/:vrn               uk.gov.hmrc.testuser.controllers
 
 GET         /organisations/empref/:empRef       uk.gov.hmrc.testuser.controllers.TestUserController.fetchOrganisationByEmpRef(empRef: uk.gov.hmrc.domain.EmpRef)
 GET         /organisations/vrn/:vrn             uk.gov.hmrc.testuser.controllers.TestUserController.fetchOrganisationByVrn(vrn: uk.gov.hmrc.domain.Vrn)
+GET         /organisations/ctUtr/:ctUtr         uk.gov.hmrc.testuser.controllers.TestUserController.fetchOrganisationByCtUtr(ctUtr: uk.gov.hmrc.domain.CtUtr)
+GET         /organisations/sautr/:saUtr         uk.gov.hmrc.testuser.controllers.TestUserController.fetchOrganisationBySaUtr(saUtr: uk.gov.hmrc.domain.SaUtr)
+GET         /organisations/crn/:crn             uk.gov.hmrc.testuser.controllers.TestUserController.fetchOrganisationByCrn(crn: uk.gov.hmrc.testuser.models.Crn)
```
Comment (ajchilts) [inline conf/app.routes:21]: Why is this in `testuser.models` and not in `domain` like the others?

**Review by ajchilts — CHANGES_REQUESTED** [review]

File: conf/app.routes (hunk 11 above)
Comment (parkeradam) [reply conf/app.routes:21]: Becase Crn is not defined in the Domain Library (or if it is I couldnt find it). As such I implemented a similar class locally.

**Pull request merged**